src/Test/Diagnostics/TestingOnly_WaitingService.cs

```diff
@@ -44,15 +56,43 @@ public void WaitForAsyncOperations(string featureName, bool waitForWorkspaceFirs
             GC.KeepAlive(featureWaiter);
         }
 
-        public void WaitForAllAsyncOperations(params string[] featureNames)
+        public ImmutableArray<KeyValuePair<string, TimeSpan>> WaitForAllAsyncOperations(params string[] featureNames)
         {
+            var longRunningTokens = new Dictionary<string, TimeSpan>();
+            var stopwatch = Stopwatch.StartNew();
```

Comment:
so, we want to know time things take after WaitForAllAsyncOperations is called? won't it be simpler if DiagnosticAsyncToken has stop watch that save start time and dispose time?

Reply:
to make the log more actionable, I think having a bit more info might be more helpful.

for example, DiagnosticAsyncToken has info such as name, file, line, tag and etc. if we add stopwatch there as well, then, this method could just return all DiagnosticAsyncToken that got finished once WaitForAllAyncOperations is done.

and the caller of this method, go through returned token to remove ones it doesn't care and log all those info above to find out exactly what point is the problem.

right now, it got aggregated for same feature and show the last one's elapsed time so no idea what is actually problem.

it doesn't tell whether it is because a specific async token or due to many async tokens queued, or one async token blocked until some other async token to be finished and etc.

Reply:
> to make the log more actionable

Would the signature of the method need to change for this, or can we continue to use a map from `string`&rarr;`TimeSpan`?

If you aren't comfortable with the current messaging, I could remove commit e12b9410f5bf7ecf9093cd500b7dd9044e753320 so the logging infrastructure is at least in place.